components/proxy/src/test/java/com/hotels/styx/admin/handlers/ServiceProviderHandlerTest.java

```diff
@@ -0,0 +1,168 @@
+import org.apache.commons.lang.StringUtils;
+import org.mockito.Mockito;
+import org.testng.annotations.Test;
+import reactor.core.publisher.Mono;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.testng.Assert.assertTrue;
+
+public class ServiceProviderHandlerTest {
+
+    @Test
+    public void returnsAllProviders() throws IOException {
+        StyxObjectStore<ProviderObjectRecord> store = createTestStore();
+        ServiceProviderHandler handler = new ServiceProviderHandler(store);
+        HttpRequest request = HttpRequest.get("/admin/service/providers").build();
+        HttpResponse response = Mono.from(handler.handle(request, HttpInterceptorContext.create())).block();
+
+        assertThat(response.status(), equalTo(HttpResponseStatus.OK));
+
+        List<StyxObjectDefinition> actualProviders = extractProviders(response.bodyAs(StandardCharsets.UTF_8));
+        assertThat(actualProviders.size(), equalTo(store.entrySet().size()));
+        for (StyxObjectDefinition actual :
+                actualProviders) {
+            Optional<ProviderObjectRecord> rec = store.get(actual.name());
+            assertTrue(rec.isPresent());
+            validateProvider(actual, rec.get());
+        }
+    }
+
+    @Test
+    public void returnsNoContentStatusWhenNoProvidersAvailable() {
+        StyxObjectStore<ProviderObjectRecord> empty = new StyxObjectStore<>();
+        ServiceProviderHandler handler = new ServiceProviderHandler(empty);
+        HttpRequest request = HttpRequest.get("/admin/service/providers").build();
+        HttpResponse response = Mono.from(handler.handle(request, HttpInterceptorContext.create())).block();
+
+        assertThat(response.status(), equalTo(HttpResponseStatus.NO_CONTENT));
+        assertThat(response.contentLength().get(), equalTo(0L));
+    }
+
+    @Test
+    public void returnsNamedProvider() throws IOException {
+        StyxObjectStore<ProviderObjectRecord> store = createTestStore();
+        ServiceProviderHandler handler = new ServiceProviderHandler(store);
+        HttpRequest request = HttpRequest.get("/admin/service/provider/object2").build();
+        HttpResponse response = Mono.from(handler.handle(request, HttpInterceptorContext.create())).block();
+
+        assertThat(response.status(), equalTo(HttpResponseStatus.OK));
+
+        StyxObjectDefinition actualProvider = deserialiseProvider(response.bodyAs(StandardCharsets.UTF_8));
+        assertThat(actualProvider.name(), equalTo("object2"));
+        assertThat(actualProvider, notNullValue());
```

Comment:
should we verify it's not null before calling .name() instead of the other way around?

Reply:
Yes, indeed.